### DISCUSSION TITLE
Questions and possible improvements for `datasource/github-runners`

### DISCUSSION THREAD

**Discussion by HonkingGoose** [renovatebot/renovate#26960] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

## Summary

Questions about our `datasource/github-runners` code due to GitHub blogging about their new MacOS 14 runners.

- How to add new MacOS 14 runner?
- Are we missing the `macos-large` variants?
- Are we using the correct names?
- Are we missing MacOS and Windows tests?

### Background info

GitHub just blogged about their new macos 14 runners. [^gh-blog] So I went looking if Renovate can "see" the old runner, and update the `runs-on:` key in the workflow files. I ended up at this code:

https://github.com/renovatebot/renovate/blob/7681ad00fb83b4d3404bfce24a695a2bf605693d/lib/modules/datasource/github-runners/index.ts#L8-L22

I checked our open issues for `github-runners` but did not find anything relevant.

## GitHub runners available images [^1]

> ## Available Images
> 
> | Image | YAML Label | Included Software |
> | --------------------|---------------------|--------------------|
> | Ubuntu 22.04 | `ubuntu-latest` or `ubuntu-22.04` | [ubuntu-22.04] |
> | Ubuntu 20.04 | `ubuntu-20.04` | [ubuntu-20.04] |
> | macOS 14 [beta] | `macos-14`, `macos-14-large`, or  `macos-14-xlarge`| [macOS-14] |
> | macOS 13 | `macos-13`, `macos-13-large`, or  `macos-13-xlarge`| [macOS-13] |
> | macOS 12 | `macos-latest`,`macos-12`, `macos-latest-large`, or `macos-12-large`| [macOS-12] |
> | macOS 11 [deprecated] | `macos-11`| [macOS-11] |
> | Windows Server 2022 | `windows-latest` or `windows-2022` | [windows-2022] |
> | Windows Server 2019 | `windows-2019` | [windows-2019] |

### How to add new MacOS 14 runner?

Would it work to add the MacOS 14 runner like this?

```diff
    macos: [
+      { version: '14' },
+      { version: '14-xl' },
+      { version: '14-l' },
      { version: '13' },
      { version: '13-xl' },
      { version: '12' },
      { version: '12-xl' },
      { version: '11' },
      { version: '10.15' },
    ],
```

### Are we missing the `macos-large` variants?

I noticed in the GitHub runners repository readme [^1] that they have a `-large` flavor for the MacOs runners. The current `datasource/github-runers-index.ts` only shows the _normal_ and _xl_ runners. So are we missing the _large_ variant in our code?

### Are we using the correct names?

Also the naming seems different:

| Our code | YAML label name |
| --- | --- |
| `13` | `macos-13` |
| missing? | `macos-13-large` |
| `13-xl` | `macos-13-xlarge` |

### Are we missing MacOS and Windows tests?

I don't see any MacOS and Windows style test in this file:

https://github.com/renovatebot/renovate/blob/7681ad00fb83b4d3404bfce24a695a2bf605693d/lib/modules/datasource/github-runners/index.spec.ts#L12-L19

[^1]: [GitHub runners repository readme, available images](https://github.com/actions/runner-images/tree/main#available-images)
[^gh-blog]: [GitHub blog, GitHub Actions: macOS 14 (Sonoma) is now available](https://github.blog/changelog/2024-01-30-github-actions-macos-14-sonoma-is-now-available/)

### Logs (if relevant)

Logs not relevant to my query.

**Comment by secustor** [comment]
> How to add new MacOS 14 runner?

Yes, exactly like that. 

> Are we missing the macos-large variants?

Yes, looks like it. 

> Are we missing MacOS and Windows tests?

You are right there are currently no test for that and there should probably ones. 
That said the logic here is very small, the biggest junk of is here: 

https://github.com/renovatebot/renovate/blob/7681ad00fb83b4d3404bfce24a695a2bf605693d/lib/modules/manager/github-actions/extract.spec.ts#L390

**Reply by HonkingGoose** [reply]
@rarkins do you want me to create issues about:

1. Adding the new macOS 14 runner, once that's stable? Nevermind, I think we should track this anyways, see issue: #27228
2. Adding the macos-large variant? See: #27229
3. Adding macOS and Windows tests?

**Edit:** turns out that the macOS 14 runner is still in beta. Source of truth is the readme in the `actions/runner-images` repository. :upside_down_face: I assume we will want to track this upgrade somewhere, so I created an issue for us (#27228).

**Reply by HonkingGoose** [reply, marked as answer]
Nevermind, @viceice is fixing all my problems with the datasource now:

- #27291
- #27292 (also adds tests 🥳 )
